lib/autoproj/ops/configuration.rb

```diff
@@ -314,57 +313,20 @@ def inspect
                 to_s
             end
 
+            # Sort the package sets by dependency order
+            # Package sets that have no dependencies come first,
+            # the local package set (by main configuration) last
             def sort_package_sets_by_import_order(package_sets, root_pkg_set)
-                # The sorting is done in two steps:
-                #  - first, we build a topological order of the package sets
-                #  - then, we insert the auto-imported packages, following this
-                #    topological order, in the user-provided order. Each package is
-                #    considered in turn, and added at the earliest place that fits
-                #    the dependencies
-                topological = Array.new
-                queue = (package_sets.to_a + [root_pkg_set]).uniq
-                until queue.empty?
-                    last_size = queue.size
-                    pending = queue.dup
-                    queue = Array.new
-                    until pending.empty?
-                        pkg_set = pending.shift
-                        if pkg_set.imports.any? { |imported_set| !topological.include?(imported_set) }
-                            queue.push(pkg_set)
-                        else
-                            topological << pkg_set
-                        end
-                    end
-                    if queue.size == last_size
-                        raise ArgumentError, "cannot resolve the dependencies between package sets. There seem to be a cycle amongst #{queue.map(&:name).sort.join(', ')}"
-                    end
-                end
-
-                result = root_pkg_set.imports.to_a.dup
-                to_insert = topological.dup
-                                       .find_all { |p| !result.include?(p) }
-                until to_insert.empty?
-                    pkg_set = to_insert.shift
-                    dependencies = pkg_set.imports.dup
-                    if dependencies.empty?
-                        result.unshift(pkg_set)
-                    else
-                        i = result.find_index do |p|
-                            dependencies.delete(p)
-                            dependencies.empty?
-                        end
-                        result.insert(i + 1, pkg_set)
-                    end
+                c = PackageSet::Comparator.new(package_sets, root_pkg_set)
+                sorted_pkg_sets = c.pkg_sets.keys.sort do |a, b|
```

Comment:
Better use `package_sets + [root_pkg_set]` here. Using `c.pkg_sets.keys` here feels like an anti-pattern to me. Or define a `all_package_sets` on Comparator and use that.